### DISCUSSION TITLE
Prior austenite reconstruction strategies

### DISCUSSION THREAD

**Discussion by nyyssont** [mtex-toolbox/mtex#569] [Ask Anything] [answered]
Hello,

I'm looking for more speed in reconstructing large martensitic datasets (about 5 million points and above). One way would be to use the austenite orientations calculated for individual misorientations across boundaries. They are rapidly calculated and on visual examination, provide consistent and in most cases accurate solutions. This is visualized in Figure1, where each boundary under a fitting value of 5 degrees has been given an IPF color corresponding to the shared solution at that boundary.

![Figure 1](https://user-images.githubusercontent.com/10060198/98570358-a81e9900-22b3-11eb-9792-b659b9957af7.jpg)

Now, each grain will have an amount of potential prior austenite orientation solutions matching the amount of neighbors. Each neighbor will share a similar solution. This is visualized in Figure 2, scenario 3, where three grains with various solutions are shown.

![Figure 2](https://user-images.githubusercontent.com/10060198/98570384-ace34d00-22b3-11eb-9d03-fe3deb187ff0.jpg)

One way to provide a guess for a correct solution for each grain would be to pick the best fitting solution (scenario 1 in attached image). The attached script does this for the example martensitic dataset and the result is shown in Figure 3. Clearly, this solution results in misindexations and leaves something to be desired.

![Figure 3](https://user-images.githubusercontent.com/10060198/98570404-b371c480-22b3-11eb-841e-cd6a3a8277fb.jpg)

Another way would be to pick the solution with the most representations in each grain (scenario 2 in Figure 2). A third way would be to pick the solution with the most representations in a cluster of grains (scenario 3).

Any suggestions on how to connect the potential solutions to each other so that the level of representation for each solution can be determined? I'm thinking some sort of modified calcGrains algorithm with the spatial decomposition bypassed somehow would be a relatively fast alternative.

I hope this explanation made sense to someone. What do you think? Any discussion on prior austenite reconstruction strategies is welcome.

[reconstruct_best_fit.zip](https://github.com/mtex-toolbox/mtex/files/5512022/reconstruct_best_fit.zip)

**Comment by ralfHielscher** [comment]
Hi Tuomo,

this looks extremely similar to what we did in [Titanium Example](https://mtex-toolbox.github.io/TiBetaReconstruction.html). There we went for 
- triple junctions instead of boundaries to be more sure about the resulting parent orientation
- required that the second best fit for a triple junction is sufficiently bad
- required all votes to be consistent.

The voting is done by the function [`majorityVote`](https://mtex-toolbox.github.io/majorityVote.html) which allows already for scenario (1) and could be easily generalized to scenario (2).

In the [Titanium Example](https://mtex-toolbox.github.io/TiBetaReconstruction.html) we proceed by

1. computing the parent orientations for each child grain that got sufficiently many reliable votes 
2. do a ```merge(grains,'threshold',5*degree)```  which simply merges similarly oriented parent grains. 
3. After we have some parent grains we can grow these into the child regions.

Ralf.

**Reply by nyyssont** [reply]
I don't know why this didn't occur to me. For some reason I felt that it is always necessary to check similarity by calculating misorientation angle, but of course this is not so when the prior austenite orientations are calculated using the same average grain orientation - then it is enough to check for the index number.

Thanks. I'll keep working on this.

**Comment by AmitS07** [comment]


**Reply by nyyssont** [reply]
I'd be happy to help you with these issues. Would you mind creating a separate topic for them? I would like to limit this thread to discussion on prior austenite reconstruction strategies.

Respectfully,
Tuomo

**Reply by AmitS07** [reply]
Dr. Tuomo,

Thank you for your kind reply.
Sure! I will do it in separate topic.

Thanks

Yours sincerely
Amit

**Comment by nyyssont** [comment]
Hello,

I've attempted to implement "Scenario 2" above in the code attached in the zip file. The similarity of the solutions within each grain are checked by connecting them all to each other and calculating the misorientation angle. The solution with the largest amount of representations is then taken as the solution for the grain. I'm sure this could be implemented using the methodology proposed by Ralf earlier, but I chose to go like this for the present.

For the demo file, some improvement can be seen compared to Scenario 1. However, apparent misindexations are still present in large amounts. Grains with a small amount of neighbors are very prone to misindexations when using this reconstruction approach.

![Scenario_2](https://user-images.githubusercontent.com/10060198/99899554-96bd8f80-2caa-11eb-96f0-5ac387c2b056.jpg)

However, this approach offers some improvement in speed and results in promising results for larger datasets. Below is a result obtained for a dataset of 2,5 million datapoints and about 170 000 martensite grains, with a runtime of 254 s on my computer. Plotting took 83 seconds of this time. Using the approach when an optimal solution is calculated separately for each cluster takes closer to a half hour.

![speed_improvement](https://user-images.githubusercontent.com/10060198/99875115-72e84400-2bed-11eb-9053-586bf58e6d87.jpg)

I will next attempt Scenario 3 and, after that, try out the approach employed for titanium reconstruction.

[reconstruct_new_logic_demo.zip](https://github.com/mtex-toolbox/mtex/files/5577400/reconstruct_new_logic_demo.zip)

**Comment by nyyssont** [comment]
The attached script implements Scenario 3 (most representations in a Markovian cluster) and the result is shown below. There is no real improvement. The grains with a low amount of neighbors or only a single neighbor are still the problem.

![Scenario_3](https://user-images.githubusercontent.com/10060198/99963004-66045580-2d91-11eb-8a9b-fde456642698.png)

Perhaps the way to go would be to not assign prior orientations to low-neighbor grains this way and simply merge them into partial prior austenite grains in a subsequent step.

Next: titanium reconstruction logic applied to martensite.

[reconstruct_new_logic_mcl_demo.zip](https://github.com/mtex-toolbox/mtex/files/5582974/reconstruct_new_logic_mcl_demo.zip)

**Comment by ralfHielscher** [comment, marked as answer]
Hi Tuomo,

could you please have a look into the branch feature/parentGrainReconstruction
I have updated the job class to include the voting algorithms. In short you can now to 
```matlab
% compute votes from grain boundaries
job.calcGBVotes

% compute parent orientations from votes - can be adapted by options - maybe you want to add more options into this function
job.calcParentFromVote('minFit',2.5*degree,'maxFit',2.5*degree,'minVotes',2)
```
The nice think is, you can iterate this very easily by simply rerunning these lines or follow it by a graph merge aka
```matlab
job.buildGraph('threshold',5*degree)
job.clusterGraph('inflationPower',2)
job.mergeByGraph('threshold',20*degree)
```
Other nice functions are `job.calcGBVotes`, `job.mergeSimilar`, `job.mergeInclusions`. What I like about this approach is that after each step all the data are in a consistent state.

I think I will soon merge this into develop and release it as mtex 5.6.beta.1 as it makes thinks so much easier and avoids many problems of the old syntax.

All the best.

Ralf.

**Reply by nyyssont** [reply]
Well this seems very handy. I'll start testing it now. Thanks Ralf.

**Comment by nyyssont** [comment]
Using the branch feature/parentGrainReconstruction, the following lines execute something like scenario 2:

```matlab
% plotting convention
setMTEXpref('xAxisDirection','east');
setMTEXpref('zAxisDirection','intoPlane');

% load the data
mtexdata martensite 

% grain reconstruction
[grains,ebsd.grainId] = calcGrains(ebsd('indexed'),'angle',3*degree);
ebsd(grains(grains.grainSize < 4)) = [];
[grains,ebsd.grainId] = calcGrains(ebsd('indexed'),'angle',3*degree);

job = parentGrainReconstructor(ebsd,grains) % define job
job.calcParent2Child; % get representative OR

job.calcGBVotes % compute votes from grain boundaries
job.calcParentFromVote('minVotes',2) % compute parent orientations from votes

job.mergeSimilar('threshold',10*degree); % Merge all similar grains
job.mergeInclusions % Merge all inclusion-type grains

figure
plot(job.grains('Iron fcc'),job.grains('Iron fcc').meanOrientation)
```

and produces the figure below:

![example](https://user-images.githubusercontent.com/10060198/101171782-8e693b00-3640-11eb-9f3c-f4d8703ba8c6.jpg)

The level of representation is determined by variant index number. This presents a few obstacles that need to be kept in mind:

1) Several martensitic variants have a very low mutual misorientation angle (can be as low as 2-4 degrees).
2) The low-angle variants generally form next to each other.
3) Consequently, quite often two variants can often be found in what MTEX interprets as "grain".

This makes the voting slightly inconsistent when grain mean orientations are used. It helps when the representative orientation relationship is as good as possible. In the case of the example, the result is quite decent and actually an almost complete reconstructed map.

I had to modify the calcParent2Child and calcParentFromVote a bit to get this result. Currently, the OR determination is not that good in my opinion (using quantile like this results in a really small sample of p2cs to average). Should I submit a pull request to feature/parentGrainReconstruction?

Now on wishlist:
1) Merge grains without determining prior orientation in parentGrainReconstructor, so the voting algorithm can be used for the merged grains (essentially implementing scenario 3 using the variant index numbers).
2) A voting algorithm that uses misorientation angles between individual prior orientation results rather than variant index numbers.

I will continue to test and tweak. In any case, I really like the new parentGrainReconstructor, it does help immensely in keeping things tidy.

EDIT: went ahead and proposed changes to feature/ParentGrainReconstruction

**Comment by nyyssont** [comment]
The more I use these new parentGrainReconstructor functions the more I like them. It's very easy to use the various tools available to reconstruct the prior structure in various stages. The script below implements the following workflow:

1. Determine representative OR
2. Determine variant index candidates for each grain based on misorientations between child orientations
3. Assign the orientation with the most variant index candidates (or "votes") to the child grain
4. Merge similar grains, merge inclusions
5. Perform two rounds of absorbing neighboring child grains to the still incomplete prior grains
6. Create graph and cluster it using MCL to fill up the remaining non-reconstructed areas, calculate prior orientations for clusters
7. Plot the results.

```matlab
% load the data
mtexdata martensite 
plotx2east

% grain reconstruction
[grains,ebsd.grainId] = calcGrains(ebsd('indexed'),'angle',3*degree);
ebsd(grains(grains.grainSize < 4)) = [];
[grains,ebsd.grainId] = calcGrains(ebsd('indexed'),'angle',3*degree);

job = parentGrainReconstructor(ebsd,grains) % define job
job.calcParent2Child; % get representative OR

job.calcGBVotes % compute votes from child grain boundaries
job.calcParentFromVote('minVotes',4) % compute parent orientations from votes

job.mergeSimilar('threshold',5*degree); % Merge all similar grains
job.mergeInclusions % Merge all inclusion-type grains

for l = 1:2
    job.calcGBVotes('onlyParents') % compute votes ONLY from parent-child boundaries
    job.calcParentFromVote('minFit',5*degree) %compute parent orientations from votes
end

job.mergeSimilar('threshold',5*degree); % Merge all similar grains

job.buildGraph('threshold',5*degree) % Make graph for clustering
job.clusterGraph('inflationPower',1.4) % Cluster using MCL
job.mergeByGraph('threshold',20*degree) % Calculate orientations for clusters

figure
plot(job.grains('Iron fcc'),job.grains('Iron fcc').meanOrientation)
```

![newfile](https://user-images.githubusercontent.com/10060198/101248253-2177a300-371d-11eb-9bbb-23858eb2ba6b.jpg)

Previously, for me at least, maintaining consistency in the indexation of the various grains, parent grains, retained austenite, votes etc. was a nightmare. Now, it looks like I don't have to worry about it at all. Note that I changed calcParentFromVote to include an option that uses only parent/child misorientations to produce votes.

Thanks, MTEX team. This is fine work and a really good foundation to build reconstruction algorithms on. Looking forward to the eventual release of parentGrainReconstructor.

**Reply by ralfHielscher** [reply]
Hi Tuomo,

I'm glad you like it. Before releasing the new job class priority for me is to fix the naming schema. Once it is released changing function and variable names is not so nice for the users as it breaks existing scripts. So if you have some suggestions here - they are welcome :)

Also you have another open pull request. I didn't merged it so far because it was to hard for me to understand. Actually this was the moment when I started the job class. So would it make sense for you to rewrite the example with the new functionality at hand.

All the best,

Ralf.

**Reply by nyyssont** [reply]
I closed down the pull request and will rewrite it for when the new job class is released!

**Comment by ralfHielscher** [comment]
Hi Tuomo,

I have now implement probability voting. Example

```matlab
job.calcGBVotes('numFit',2)

job.calcParentFromVote('probability', 'threshold',2*degree,'tolerance',1.5*degree,'minProp',0.55)
```
This computes from the misfit a number between 0 and 1 with the same function as for the similarity matrix that is controlled by the parameters threshold and tolerance. 1 - means perfect fit - 0  means no fit. Then these probabilities are summed up for the different votes and the vote with the maximum value is considered. Also the second best probability is computed and the ratio prob1 / (prob1 + prob2) is compared against minProp. Ratio close to 1 means we are very sure about this vote. Ratio close to 0.5 means we have two equally good solutions.

Would you mind testing this?

Ralf.

**Reply by nyyssont** [reply]
Doing:

```matlab
% load the data
mtexdata martensite 
plotx2east

% grain reconstruction
[grains,ebsd.grainId] = calcGrains(ebsd('indexed'),'angle',3*degree);
ebsd(grains(grains.grainSize < 4)) = [];
[grains,ebsd.grainId] = calcGrains(ebsd('indexed'),'angle',3*degree);

job = parentGrainReconstructor(ebsd,grains) % define job
job.calcParent2Child; % get representative OR

job.calcGBVotes('numFit',2)
job.calcParentFromVote('probability', 'threshold',2*degree,'tolerance',1.5*degree,'minProp',0.55)
job.mergeSimilar('threshold',5*degree); % Merge all similar grains

figure
plot(job.grains('Iron fcc'),job.grains('Iron fcc').meanOrientation)
```

Resulted in:

![rec_result](https://user-images.githubusercontent.com/10060198/102690057-1b94be00-4203-11eb-8856-d0f68a2867e2.jpg)

This is an excellent and really fast result with minimal effort and no editing of your code, although I tweaked the calcParent2Child script to get a (in my opinion) more representative OR. I will have to do later testing in January, as I'm now heading for vacation while the computer stays at home. But this seems to be working well and I will look at the code more closely next year.

**Reply by nyyssont** [reply]
I'd say the 'minProp' looks really promising to get more precision for twin boundaries - though I don't have a concrete suggestion right away how to implement it for that purpose.

**Reply by ralfHielscher** [reply]
important: 'minProp' was a spelling mistake. It should be 'minProb'

Ralf.

**Reply by ralfHielscher** [reply]
Also, it makes live a bit easier and reliable to use `job.parentGrains` instead of `job.grains('Iron fcc')`

**Comment by ralfHielscher** [comment]
Hi Tuomo,

I have set up the project https://github.com/mtex-toolbox/mtex/projects/5 to keep a better overview about our activities. 
If you have a feature request do the following:

1. open an issue with this feature request
2. assign this issue to the project

There was one more feature request above that I was not able to understand:

> Merge grains without determining prior orientation in parentGrainReconstructor, so the voting algorithm can be used for the merged grains (essentially implementing scenario 3 using the variant index numbers).

I have also added a new feature to undo a reconstruction. The idea is to identify wrongly reconstructed grains at put them into their original status and then try again :smiley:

Ralf.
